refactor(executor): experimental support of cancellable query execution.

arkbriar opened the pull request:
I'm going to close #422 because it's far more complicated to interrupt a query and make things right when a query is cancelled, such as aborting/rolling back transactions and cleaning up spawned tasks.

This PR is not only to #423. Here're the main contributions of this PR:

1. Use cancellation token to signal and detect the cancellation.
2. Provides a non-invasive interface `cancellable` to provide the cancellation of executors and integrates it inside the executor tree. 
  + IMO, only the leaf nodes with much IO works should be sensitive to the cancellation.
  + Resources such as transactions are rollbacked in a spawned task. We think spawned tasks as completable tasks as long as the async runtime is still running. This avoids the complex vacuum thread and cleans the room ASAP.
3. Provide a struct and a set of interfaces to organize the spawned tasks during execution, which avoids leaks while shutting down the async runtime (exit process immediately after interrupting the query).

Here I only include the cancellation of `TableScanExecutor` in this PR. However, the others (e.g. insert/delete/copy) can be supported in the same fashion.

Now you can interrupt a select query in the CLI at any time as you type the CTRL-C. The transactions shall be dealt properly.

To show the debug messages below, please run with `RUST_LOG=risinglight=debug cargo run --release`.

```bash
RUST_LOG=risinglight=debug cargo run --release
> select
    l_orderkey,
    sum(l_extendedprice * (1 - l_discount)) as revenue,
    o_orderdate,
    o_shippriority
from
    customer,
    orders,
    lineitem
where
    c_mktsegment = 'BUILDING'
    and c_custkey = o_custkey
    and l_orderkey = o_orderkey
    and o_orderdate < date '1995-03-15'
    and l_shipdate > date '1995-03-15'
group by
    l_orderkey,
    o_orderdate,
    o_shippriority
order by
    revenue desc,
    o_orderdate
limit 10;

...
^CInterrupted
2022-02-14T15:15:20.381454Z  WARN risinglight::executor::table_scan: Abort!
2022-02-14T15:15:20.381506Z DEBUG risinglight::db: error: abort
```

TennyZhuang commented on Cargo.toml:
```diff
@@ -50,6 +50,7 @@ sqlparser = { git = "https://github.com/risinglightdb/sqlparser-rs", rev = "edea
 thiserror = "1"
 tikv-jemallocator = { version = "0.4", optional = true }
 tokio = { version = "1", features = ["full"] }
+tokio-util = "0.7.0"
```
```suggestion
tokio-util = "0.7"
```

TennyZhuang commented on src/executor/mod.rs:
```diff
@@ -114,20 +120,89 @@ pub type BoxedExecutor = BoxStream<'static, Result<DataChunk, ExecutorError>>;
 
 impl ExecutorBuilder {
     /// Create a new executor builder.
     pub fn new(storage: StorageImpl) -> ExecutorBuilder {
```
Do we even need the method now?

arkbriar replied:
Not really, just for backward compatibility. Remove it should be nicer.

TennyZhuang reviewed (approved):
Great work! Generally LGTM, but it seems that we may need some more comments.

TennyZhuang commented:
And please resolve conflicts.

arkbriar commented:
> Great work! Generally LGTM, but it seems that we may need some more comments.

Thanks @TennyZhuang . I am greatly encouraged.

I added more commends/docs on `SealableAtomicCounter` and `WaitGroup` and created a new package called `utils` and moved the `sync` package to be under the `utils`. 

Also, 4 more executors (insert/delete, copy from/to file) are refactored and they are all cancellable now. The leaf executors are all cancellable and I think we can interrupt any query from now on.

> And please resolve conflicts.

All conflicts should be resolved now.

Please review this again when you have time.

TennyZhuang commented:
Thanks! I think it’s good enough to merge!
